lib/Common/Memory/LargeHeapBlock.cpp

```diff
@@ -280,60 +286,112 @@ LargeHeapBlock::ReleasePagesShutdown(Recycler * recycler)
+        {
+            uint64 current = alignAddr[i];
+            if (current != 0xF0F0F0F0F0F0F0F0u)
+            {
+                Assert(false);
+                ReportFatalException(NULL, E_FAIL, Fatal_Recycler_MemoryCorruption, 2);
+            }
+        }
+#else
+        byte* addr = (byte*)this->allocAddressEnd;
+        uint32* alignAddr = (uint32*)Math::Align(((uint32)addr), sizeof(void*));
+
+        for (int i = 0; addr + i<(byte*)alignAddr; i++)
+        {
+            byte current = addr[i];
+            if (current != 0xF0u)
+            {
+                Assert(false);
+                ReportFatalException(NULL, E_FAIL, Fatal_Recycler_MemoryCorruption, 2);
+            }
+        }
 
-template void LargeHeapBlock::ReleasePages<true>(Recycler * recycler);
-template void LargeHeapBlock::ReleasePages<false>(Recycler * recycler);
+        for (int i = 0; alignAddr + i < (uint32*)this->allocAddressEnd; i++)
+        {
+            uint32 current = alignAddr[i];
+            if (current != 0xF0F0F0F0u)
+            {
+                Assert(false);
+                ReportFatalException(NULL, E_FAIL, Fatal_Recycler_MemoryCorruption, 2);
+            }
+        }
+#endif
+    }
+}
+#endif
 
-template<bool pageheap>
 void
 LargeHeapBlock::ReleasePages(Recycler * recycler)
 {
     Assert(segment != nullptr);
 
-    char* pageAddress = address;
-    size_t realPageCount = pageCount;
+    char* blockStartAddress = this->address;
+    size_t realPageCount = this->pageCount;
 #ifdef RECYCLER_PAGE_HEAP
-    if (pageheap)
+    if (InPageHeapMode())
     {
-        if (InPageHeapMode())
-        {
-            if (guardPageAddress != nullptr)
-            {
-                DWORD noAccess;
-                if (::VirtualProtect(static_cast<LPVOID>(guardPageAddress), AutoSystemInfo::PageSize, guardPageOldProtectFlags, &noAccess) == FALSE)
-                {
-                    AssertMsg(false, "Unable to set permission for guard page.");
-                    return;
-                }
-                AssertMsg(noAccess == PAGE_NOACCESS, "Guard page should be PAGE_NOACCESS");
+        Assert(((LargeObjectHeader*)this->address)->isPageHeapFillVerified || this->allocCount == 0);
+        VerifyPageHeapPattern();
 
-                if (this->pageHeapMode == PageHeapMode::PageHeapModeBlockStart)
-                {
-                    pageAddress = guardPageAddress;
-                }
+        if (guardPageAddress != nullptr)
+        {
 
-                realPageCount = actualPageCount;
+            if (this->pageHeapMode == PageHeapMode::PageHeapModeBlockStart)
+            {
+                blockStartAddress = guardPageAddress;
             }
+            realPageCount = actualPageCount;
+
+            MEMORY_BASIC_INFORMATION memInfo;
+            VirtualQuery(this->address, &memInfo, sizeof(memInfo));
+            if (::VirtualAlloc(guardPageAddress, AutoSystemInfo::PageSize, MEM_COMMIT, memInfo.Protect) == NULL)
```

Comment:
this is for page reusing. I think we should implement a mechanism in pageAllocator to only commit part of the pages, like when asking 2 pages but only commit first page for me. and let the pageAllocator to manage the commit status. 
with that we don't even need to decommit while pageheap allocating. that is also good for some of the medium block, which last 1 or more page is not allocatable, we ask for 8 pages but only commit first 7 pages.

---
In reply to: [60994617](https://github.com/Microsoft/ChakraCore/pull/711#discussion_r60994617) [](ancestors = 60994617)